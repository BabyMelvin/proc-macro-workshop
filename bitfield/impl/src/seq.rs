use proc_macro2::{TokenStream as TokenStream2, TokenTree as TT};
use syn::{buffer::Cursor, Ident, LitInt, Token};

#[allow(dead_code)]
pub struct Seq {
    ident:       Ident,
    in_token:    Token![in],
    lhs:         LitInt,
    dot2_token:  Token![..],
    eq_token:    Token![=],
    rhs:         LitInt,
    brace_token: syn::token::Brace,
    pub tokens:  TokenStream2,
}

impl syn::parse::Parse for Seq {
    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
        let ident = input.parse()?;
        let in_token = input.parse()?;
        let lhs: LitInt = input.parse()?;
        let dot2_token = input.parse()?;
        let eq_token = input.parse()?;
        let rhs: LitInt = input.parse()?;

        let content;
        let brace_token = syn::braced!(content in input);

        let range = match (lhs.base10_parse(), rhs.base10_parse()) {
            (Ok(l), Ok(r)) => l..=r,
            _ => unreachable!("请输入 u8 范围"),
        };

        let cursor = content.cursor();
        let tokens = range.map(|lit| process(cursor, &ident, lit)).collect();

        Ok(Seq { ident,
                 in_token,
                 lhs,
                 dot2_token,
                 eq_token,
                 rhs,
                 brace_token,
                 tokens })
    }
}

fn process(mut cursor: Cursor, ident: &Ident, lit: u8) -> TokenStream2 {
    let mut ts: Vec<TT> = Vec::with_capacity(8);

    while let Some((token, cur)) = cursor.token_tree() {
        cursor = cur;
        match token {
            TT::Ident(i) => {
                if &i == ident {
                    ts.push(proc_macro2::Literal::u8_unsuffixed(lit).into());
                } else if let Some(c) = search_tidle_lit(cur, ident) {
                    cursor = c;
                    ts.push(quote::format_ident!("{}{}", i, lit).into());
                } else {
                    ts.push(i.into())
                }
            }
            TT::Group(ref g) => {
                let buf = syn::buffer::TokenBuffer::new2(g.stream());
                let mut group = proc_macro2::Group::new(g.delimiter(), process(buf.begin(), ident, lit));
                group.set_span(g.span());
                ts.push(group.into());
            }
            t => ts.push(t),
        }
    }

    dbg!(ts.into_iter().collect())
}

fn search_tidle_lit<'c>(cur: Cursor<'c>, ident: &Ident) -> Option<Cursor<'c>> {
    cur.token_tree().and_then(|(token, c)| match token {
                        TT::Punct(p) if p.as_char() == '~' => {
                            c.ident().and_then(|(i, c)| if &i == ident { Some(c) } else { None })
                        }
                        _ => None,
                    })
}
