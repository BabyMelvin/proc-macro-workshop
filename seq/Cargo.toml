[package]
name = "seq"
version = "0.0.0"
edition = "2021"
autotests = false
publish = false

[lib]
proc-macro = true

[[test]]
name = "tests"
path = "tests/progress.rs"

[dev-dependencies]
trybuild = { version = "1.0.49", features = ["diff"] }

[dependencies]
quote = "1"
# syn = "1"
syn = {version = "1", features = ["extra-traits"]}
proc-macro2 = "1"

# [[bin]]
# path = "./tests/02-parse-body.rs"
# name = "02"
#
# [[bin]]
# path = "./tests/04-paste-ident.rs"
# name = "04"
#
# [[bin]]
# path = "./tests/04-paste-ident2.rs"
# name = "042"
#
# [[bin]]
# path = "./tests/05-repeat-section.rs"
# name = "05"
#
# [[bin]]
# path = "./tests/05-repeat-section2.rs"
# name = "052"
#
# [[bin]]
# path = "./tests/06-init-array.rs"
# name = "06"
#
# [[bin]]
# path = "./tests/07-inclusive-range.rs"
# name = "07"
#
# [[bin]]
# path = "./tests/09-interaction-with-macrorules.rs"
# name = "09"
