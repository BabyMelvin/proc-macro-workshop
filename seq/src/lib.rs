#![allow(unused)]

use proc_macro::TokenStream;
use proc_macro2::{Group, Ident, Literal, TokenStream as TokenStream2, TokenTree};
use quote::quote;
use syn::{parse::Parse, LitInt, Token};

#[proc_macro]
pub fn seq(input: TokenStream) -> TokenStream {
    let seq = syn::parse_macro_input!(input as Seq);
    TokenStream::from(seq.expan())
}

struct Seq {
    ident:       Ident,
    in_token:    Token![in],
    lhs:         LitInt,
    dot2_token:  Token![..],
    rhs:         LitInt,
    brace_token: syn::token::Brace,
    tokens:      TokenStream2,
}

impl Parse for Seq {
    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
        let content;
        Ok(Seq { ident:       input.parse()?,
                 in_token:    input.parse()?,
                 lhs:         input.parse()?,
                 dot2_token:  input.parse()?,
                 rhs:         input.parse()?,
                 brace_token: syn::braced!(content in input),
                 tokens:      content.parse()?, })
    }
}

impl Seq {
    fn expan(self) -> TokenStream2 {
        let Seq { ident, lhs, rhs, tokens, .. } = self;
        let buffer = syn::buffer::TokenBuffer::new2(tokens);
        let cursor = buffer.begin();
        let range = lhs.base10_parse::<usize>().unwrap()..rhs.base10_parse::<usize>().unwrap();
        repeat::SeqToken::new(cursor, &ident, range).token_stream()
    }
}

mod repeat;
mod replace;

// 把 Group 内的 TokenStream 替换掉（保留 delimiter 和 span）
fn new_group(g: &Group, ts: TokenStream2) -> Group {
    let mut group = Group::new(g.delimiter(), ts);
    group.set_span(g.span());
    group
}
