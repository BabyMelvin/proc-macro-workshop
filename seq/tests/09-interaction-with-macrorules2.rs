//! 对照组：如果不使用 `09-interaction-with-macrorules.rs` 的技巧，那么无法成功生成 PROCS
//!
//! 因为声明宏生成的常量不一定在过程宏展开时被知晓（也可能那个常量根本还未生成）。

use seq::seq;

macro_rules! literal_identity_macro {
    () => {
        256
    };
}

// Expands to: `const NPROC: usize = 256;`
const NPROC: usize = literal_identity_macro!();

struct Proc;

impl Proc {
    const fn new() -> Self { Proc }
}

// Expands to: `static PROCS: [Proc; NPROC] = [Proc::new(), ..., Proc::new()];`
static PROCS: [Proc; NPROC] = seq!(N in 0..NPROC { [#(Proc::new(),)*] });

fn main() {}
