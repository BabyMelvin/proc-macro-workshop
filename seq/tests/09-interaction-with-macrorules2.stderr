error: expected integer literal
  --> tests/09-interaction-with-macrorules2.rs:23:44
   |
23 | static PROCS: [Proc; NPROC] = seq!(N in 0..NPROC { [#(Proc::new(),)*] });
   |                                            ^^^^^
